src/rust.rs

```diff
@@ -249,6 +249,15 @@ impl Drop for JSEngine {
     }
 }
 
+pub fn to_source_text<T, U>(source: &[T]) -> SourceText<U> {
+    SourceText {
+        units_: source.as_ptr() as *const _,
+        length_: source.len() as u32,
```

Comment:
@nox after trying to use the `&str` as argument version in Servo, I found it's not so convenient to pass a string slice because we might want to `encode_utf16` first.

Then, I tried to change it to `fn to_source_text<T>(source: &[T]) -> SourceText<T>`; however, while passing `&[u8]`, we can't return a `SourceText<u8>`. I think that's what you were talking about we can't just pass the wrapper into `Evaluate2`.

So, I changed the type to current version 🤔 but I wonder, is it possible to have some type level restriction to avoid some misusing here?

Or, I'm wondering is it possible or better to define `impl From` for `SourceText` 🤔?

Reply:
I would just make two functions, returning 2 different `SourceText<_>`.

Reply:
oh! it sounds good to me so that we can restrict the type for `u8` and `u16`, thanks!

Reply:
The entry for the UTF-8 one should be a `&str`, not an `&[u8]`. And the output should be a `SourceText<mozilla::Utf8Unit>`.

Reply:
Updated to fix! Should be better now 👀 Thanks!